src/librustc/ty/print/pretty.rs

```diff
@@ -1532,12 +1536,103 @@ define_print_and_forward_display! {
     }
 
     &'tcx ty::Const<'tcx> {
-        match self.val {
-            ConstValue::Unevaluated(..) |
-            ConstValue::Infer(..) => p!(write("_")),
-            ConstValue::Param(ParamConst { name, .. }) => p!(write("{}", name)),
-            _ => p!(write("{:?}", self)),
+        let u8 = cx.tcx().types.u8;
+        if let ty::FnDef(did, substs) = self.ty.sty {
+            p!(print_value_path(did, substs));
+            return Ok(cx);
+        }
+        if let ConstValue::Unevaluated(did, substs) = self.val {
+            match cx.tcx().describe_def(did) {
+                | Some(Def::Static(_))
+                | Some(Def::Const(_))
+                | Some(Def::AssociatedConst(_)) => p!(print_value_path(did, substs)),
+                _ => if did.is_local() {
+                    let span = cx.tcx().def_span(did);
+                    if let Ok(snip) = cx.tcx().sess.source_map().span_to_snippet(span) {
+                        p!(write("{}", snip))
+                    } else {
+                        p!(write("_"))
+                    }
+                } else {
+                    p!(write("_"))
+                },
+            }
+            return Ok(cx);
```

Comment:
Why not print the type for this?

Reply:
We may end up with `_: usize` for array lengths and similar things. I thought just `_` would be better

Reply:
You might want to special-case arrays not to print the type, but I'm not sure how.
Maybe a extra argument to `print_const`?